src/mutable_list.jl

```diff
@@ -2,71 +2,95 @@ mutable struct ListNode{T}
-        l.len = 0
-        l.node = ListNode{T}()
-        l.node.next = l.node
-        l.node.prev = l.node
-        return l
+    function MutableLinkedList{T}() where {T}
+        return new{T}(0, ListNode{T}())
     end
 end
 
 MutableLinkedList() = MutableLinkedList{Any}()
-
-function MutableLinkedList{T}(elts...) where T
-    l = MutableLinkedList{T}()
-    for elt in elts
-        push!(l, elt)
-    end
-    return l
-end
+MutableLinkedList(elts...) = MutableLinkedList{eltype(elts)}(elts...)
+MutableLinkedList{T}(elts...) where {T} = append!(MutableLinkedList{T}(), elts)
 
 Base.iterate(l::MutableLinkedList) = l.len == 0 ? nothing : (l.node.next.data, l.node.next.next)
 Base.iterate(l::MutableLinkedList, n::ListNode) = n === l.node ? nothing : (n.data, n.next)
 
 Base.isempty(l::MutableLinkedList) = l.len == 0
 Base.length(l::MutableLinkedList) = l.len
-Base.collect(l::MutableLinkedList{T}) where T = T[x for x in l]
-Base.eltype(::Type{<:MutableLinkedList{T}}) where T = T
+Base.collect(l::MutableLinkedList{T}) where {T} = T[x for x in l]
+Base.eltype(::Type{<:MutableLinkedList{T}}) where {T} = T
 Base.lastindex(l::MutableLinkedList) = l.len
 
+_boundscheck(l, idx) = 0 < idx <= l.len || throw(BoundsError(l, idx))
+
+# mend two nodes together
+nconnect(a::ListNode{T}, b::ListNode{T}) where {T} = a.next, b.prev = b, a
+
+# traverse to idx either from front or back, depending on whats faster
+function ntraverse(l::MutableLinkedList, idx::Int)
+    n = length(l)
+    node = l.node
+    if idx < n ÷ 2
+        for _ in 1:idx
+            node = node.next
+        end
+    else
+        for _ in 1:(n-idx+1)
+            node = node.prev
+        end
+    end
+    return node
+end
+
+# remove node from the list, the node itself is unchanged
+function nremove(l::MutableLinkedList, node::ListNode)
+    nconnect(node.prev, node.next)
+    l.len -= 1
+    return node
+end
+
+# create a new node, insert it after the provided `node` and return it
+function ninsert(l::MutableLinkedList{T}, node::ListNode{T}, data) where {T}
+    ins = ListNode{T}(node, data, node.next)
+    l.len += 1
+    return ins
+end
+
 function Base.first(l::MutableLinkedList)
-    isempty(l) && throw(ArgumentError("List is empty"))
+    _boundscheck(l, 1)
     return l.node.next.data
 end
 
 function Base.last(l::MutableLinkedList)
-    isempty(l) && throw(ArgumentError("List is empty"))
+    _boundscheck(l, length(l))
     return l.node.prev.data
 end
 
 Base.:(==)(l1::MutableLinkedList{T}, l2::MutableLinkedList{S}) where {T,S} = false
 
-function Base.:(==)(l1::MutableLinkedList{T}, l2::MutableLinkedList{T}) where T
+function Base.:(==)(l1::MutableLinkedList{T}, l2::MutableLinkedList{T}) where {T}
     length(l1) == length(l2) || return false
     for (i, j) in zip(l1, l2)
         i == j || return false
     end
     return true
 end
 
-function Base.map(f::Base.Callable, l::MutableLinkedList{T}) where T
+function Base.map(f::Base.Callable, l::MutableLinkedList{T}) where {T}
```

Comment:
It would be good to strip out a lot of these constraints that do not follow modern practices.

```suggestion
function Base.map(f, l::MutableLinkedList{T}) where {T}
```